timemachine/fe/single_topology.py

```diff
@@ -1604,13 +1594,56 @@ def _align_and_interpolate_bonded_term(self, lamb, src_potential, dst_potential,
             dst_potential.potential.idxs,
             dst_potential.params,
         )
+
+        # (ytz): sigh, this is the garbage code that we need to write if we don't do re-shaping
+        # at the potential level, sigh. If we have zero length arrays, for both src_potential and dst_potential,
+        # then we lose information about the correct shapes.
```

Comment:
nit: I'm a bit confused by this comment: I don't think we necessarily lose information about the correct shapes for zero-length arrays (e.g. an empty array of torsion indices should have shape `(0, 4)`)? Is the issue that there are still places in the code where we construct parameter arrays from lists without specifying the desired shape? Would this be addressed by adding an assertion to `BoundPotential.__post_init__`?

Reply:
Sorry this was probably an overreaction on my part. To be more precise: we do have *shape* information defined on :

`src_potential.potential.idxs, src_potential.params, dst_potential.potential.idxs, dst_potential.params`

However, if they're of length zero, the following loop:

``` python
bonded_idxs = []
bonded_params = []
for idxs, src_params, dst_params in set_of_tuples:
    bonded_idxs.append(idxs)
    bonded_params.append(interpolate_fn(idxs, src_params, dst_params, lamb))
```

Will result in length 0 lists for `bonded_idxs`, and `bonded_params`

Then I need to re-determine the `ndim` of the original inputs (or write a conditionals based on type of input), and if it's 1 or 2, check `shape[-1]`, re-shape the `bond_idxs` and `bond_params`:

``` python
if src_params_ndim == 1:
    # chiral atom
    bonded_params = jnp.array(bonded_params, dtype=src_params_dtype)
    pass
elif src_params_ndim == 2:
    # everything else
    src_params_last_shape = src_potential.params.shape[1]
    bonded_params = jnp.array(bonded_params, dtype=src_params_dtype).reshape(-1, src_params_last_shape)
else:
    assert 0
```

Instead, if each `Potential` class automatically knows how to re-shape on its own, then I can just pass in the `bond_idxs=[]` etc. and rely on the class's `__post_init__()` to do the re-shaping.

Reply:
I think the worst is that this pattern is repeated in *many* parts of the code, any time there's a risk of a zero length array/list this type of pattern pops-up.

Reply:
> Instead, if each Potential class automatically knows how to re-shape on its own, then I can just pass in the HarmonicBond(idxs=np.arrray([])) # wrong shape and dtype etc. and rely on the class's __post_init__() to do the re-shaping.

I'm probably misunderstanding something, but I don't see why the conditional branches on `ndim` would be necessary in practice? Wouldn't you typically do something like

``` python
idxs = np.array(possibly_empty_list_of_pairs, dtype=np.int32).reshape(-1, 2)
HarmonicBond(idxs)
```
?

I'm not completely against handling the conversion in `__post_init__`, but it does seem like there's a potential for surprising behavior if the input array is not empty. Maybe a compromise where we convert empty input arrays to the required shape and dtype, but assert that non-empty arrays already have the correct shape and dtype, would be palatable to everyone?

Reply:
> I'm probably misunderstanding something, but I don't see why the conditional branches on ndim would be necessary in practice? Wouldn't you typically do something like

`idxs` are fine, as they always have `ndims=2`, but `params` are variadiac.

eg: HarmonicBonds:
``` python
params_for_bonds = np.array(possibly_empty_list_of_pairs, dtype=np.float64).reshape(-1, 2)
HarmonicBond(...).bind(params_for_bonds)
```

but for ChiralAtomRestraints and ChiralBdonRestraints, `ndims=1`:
``` python
params_for_chiral_atoms = np.array(possibly_empty_list_of_pairs, dtype=np.float64).reshape(-1)
ChiralAtomRestraints(...).bind(params_for_chiral_atoms )
```

Reply:
(let's defer this to a future design session) - I'm okay with things as-is for now.